### PR TITLE
use fsync to force persisting the meta file.

### PR DESCRIPTION
**Pull request by krasi-georgiev**
fixes https://github.com/prometheus/prometheus/issues/4058

Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

File: block.go [hunk 1]
```diff
@@ -251,6 +251,10 @@ func writeMetaFile(dir string, meta *BlockMeta) error {
 		merr.Add(f.Close())
 		return merr.Err()
 	}
+	// Force the kernel to persist the file on disk to avoid data loss if the host crashes.
+	if err := f.Sync(); err != nil {
```
Comment (pborzenkov) [inline block.go]: Hmm. Shouldn't you close the file in this case to avoid fd leak?

Comment (krasi-georgiev) [reply block.go]: yep! updated

Comment (pborzenkov) [reply block.go]: There is also a tsdb.fileutil.Sync() function which issues correct sync on macOS (default fsync doesn't issue disk barrier). It's probably better to use it.
This has been fixed in Go 1.12, so if Prometheus/TSDB require at least 1.12 it's not a big deal.

Comment (krasi-georgiev) [reply block.go]: yeah I noticed this and some parts in the code use fileutil.Fsync and others use the regular one.
I will check if there is any blocker to require 1.12 for tsdb/Prometheus and will probably open another PR to remove the workaround.

File: block.go [hunk 2]
```diff
@@ -251,6 +251,11 @@ func writeMetaFile(dir string, meta *BlockMeta) error {
...
```
Comment (gouthamve) [inline block.go]: We should also do `merr.Add(err)` so that we don't drop the error from fsync.

**Pull request merged**